### ISSUE TITLE
Fatal error using osgi platform 

### ISSUE THREAD

**Issue by scarlet033** [zgmnkv/zwave4j#36]
Hi,
I am using the zwave4j in a simple java project to control my z-stick controller, the program work well and I receive all notifications normally.
But when I move the same code into an eclipse project running on OSGI platform, the program crash with this error :
----------
A fatal error has been detected by the Java Runtime Environment:

  SIGSEGV (0xb) at pc=0x00007f87208408e4, pid=3949, tid=140218030425856

JRE version: Java(TM) SE Runtime Environment (7.0_79-b15) (build 1.7.0_79-b15)
 Java VM: Java HotSpot(TM) 64-Bit Server VM (24.79-b02 mixed mode linux-amd64 compressed oops)
Problematic frame:
V  [libjvm.so+0x63e8e4]  jni_GetStaticFieldID+0xe4

 Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again

 An error report file with more information is saved as:
 project_path/target/products/mn-cse/linux/gtk/x86_64/hs_err_pid3949.log

 If you would like to submit a bug report, please visit:
   http://bugreport.java.com/bugreport/crash.jsp
---------

I tried "ulimit -c unlimited", I got a similar error
Any Idea about the problem ? 

Thanks

**Comment by aygalinc** [comment]
Hello,

I have the same issue. Do you solved it ? 

Regards,

**Comment by aygalinc** [comment]
I am using the 0.3 official release but i see that you have added default case in your switch case that would fix this issue. But this change aren't published in a maven central. Can you make a release ?

**Comment by Neuronix2** [comment]
Take fixed library from here: https://github.com/Neuronix2/IRISv2/tree/master/iris-core/lib
You can add it into POM simply like on this code:

`        <dependency>
            <groupId>com.github.zgmnkv</groupId>
            <artifactId>zwave4j</artifactId>
            <version>0.3-SNAPSHOT</version>
            <scope>system</scope>
            <systemPath>${basedir}/../iris-core/lib/zwave4j-0.3-SNAPSHOT.jar</systemPath>
        </dependency>`

**Comment by aygalinc** [comment]
Thanks !

**Comment by aygalinc** [comment]
The problem persist but i find the bug.

If it works well on plain Java but not on OSGi, i track a classloader issue.
And in init.cpp, the classloader used to load the java class is the currentclassloader of the Thread.
So before load the library, you must set your ThreadCurrentClassLoader ( Thread.currentThread().setContextClassLoader(NativeLibraryLoader.class.getClassLoader()); for instance) with a classloader that contains the library java classes.

I think if the library used directly the classloader of his own class with avoid this.

Regards,

**Comment by mbenalaya** [comment]
Hello Aygalinc,

I think you are right. However, I was not able to test your solution. Could you please give me more details ?
Thank you in advance,
Best regards

**Comment by aygalinc** [comment]
in init.cpp :
`JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM * vm, void * pvt)
{
	jvm = vm;
	JNIEnv * env;
	if (jvm->GetEnv((void **) &env, JNI_VERSION_1_6))
	{
         return JNI_ERR; /* JNI version not supported */
    }

    jclass threadClass = env->FindClass("java/lang/Thread");
    classLoader = env->NewGlobalRef(env->CallObjectMethod(
            env->CallStaticObjectMethod(
                    threadClass,
                    env->GetStaticMethodID(threadClass, "currentThread", "()Ljava/lang/Thread;")
            ),
            env->GetMethodID(threadClass, "getContextClassLoader", "()Ljava/lang/ClassLoader;")
    ));

	return JNI_VERSION_1_6;
}`

It uses the ClassLoader of the thread calling the  `NativeLibraryLoader.loadLibrary(arg1,arg2);` method. But we can't ensure in OSGi that the classloader of the thread calling this method is the one that have load the library. 

So i recommand to use the library as following in OSGi:
`
Thread.currentThread().setContextClassLoader(NativeLibraryLoader.class.getClassLoader());
NativeLibraryLoader.loadLibrary(ZWave4j.LIBRARY_NAME, ZWave4j.class);
Options options = Options.create("config/", "", "");
options.lock();
manager = Manager.create();
manager.addWatcher(watcher, null);
....
`
It will fix the current thread classloader to the one that have loaded the library.
Does it answered your question ? 

Regards,
Colin

**Comment by mbenalaya** [comment]
Hello Colin,

Thank for this detailed answer. I really appreciate.
It should be ok. I will test it and get back to you soon.

Regards,
Mahdi

**Comment by zgmnkv** [comment]
I have commited a possible fix for this issue, can you please check it?

**Comment by aygalinc** [comment]
I can test, but your snapshot repo link is dead. Do you have a new one ?

**Comment by zgmnkv** [comment]
Can you build it yourself? Or if you tell your os/architecture, I can build it for you. Currently it's hard for me to build every rev for all os/architectures.

**Comment by aygalinc** [comment]
If you can build for a linux 64 bits it will be fine ?

Sorry to ask for this but i'm not a gradle expert.

**Comment by zgmnkv** [comment]
I'll be able to build it on the next week

**Comment by zgmnkv** [comment]
I have uploded a snapshot with linux 64 bit binary only https://oss.sonatype.org/content/repositories/snapshots/com/github/zgmnkv/zwave4j/

**Comment by scarlet033** [comment]
Hi,
Finally I had time to test your solutions,
I downloaded the jar file uploaded by zgmnkv for linux 64 bit, It resolves the problem and I receive normally notifications in an OSGI platform.
I tried also to add this line "Thread.currentThread().setContextClassLoader(NativeLibraryLoader.class.getClassLoader());" before loading the library as mentioned by Colin, I got the same error.
Thank you Colin to reopen this discussion :), and Thank you zgmnkv for the solution.

Regards 
Scarlet
